### ISSUE TITLE
Support for snapshot testing

### ISSUE THREAD

**Issue by vlsi** [kotest/kotest#395]
Consider the following test case:

Given a SQL query the system should produce the following results (in any order).
The case is repeated for various SQL queries.

Of course I can use

```
sql = "..."
assertEquals("results for ${sql}", """
   expected line 1
   expected line 2
""",  execute(sql).sorted())
sql = "..."
assertEquals("results for ${sql}", """
   expected line 1
   expected line 2
""",  execute(sql).sorted())
sql = "..."
assertEquals("results for ${sql}", """
   expected line 1
   expected line 2
""",  execute(sql).sorted())
```

however if the implementation changes, then it would be very hard to update all the test to new expected values.
It would basically force the developer to manually go and update each and every `assert` statement.

I wonder if `kotlintest` could integrate something like "self-reproducing" test cases, so it creates "new test code" as one of the results of the test execution.

That is:
1) It performs the assertions as usual
2) It writes "target/tmp/...kt" file with new source code. If developer agrees that all the "failures" are expected, he can just grab the new file and copy it over the existing one

Here's an example of such a test: https://github.com/apache/calcite/blob/3c6b5ec759caadabb67f09d7a4963cc7d9386d0c/core/src/test/resources/sql/join.iq

For instance, `!plan` at line 62 asserts that execution plan of a previous query matches the plan listed at lines 39..61.
The result of `.iq` execution is a brand new `.iq` file with "expected values" set to the current actuals.

In other words, if a test passes, then `.iq` is a quine.

`.iq` stands for https://github.com/julianhyde/quidem there.

I wonder if that test idiom can be incorporated and expressed in Kotlin.

Pros are as follows:

1) The test is somewhat easy to read
2) The test is easy to write: just start with EMPTY assertion, and test output would produce the text of the test WITH expected value for you
3) It is trivial to maintain: one reviews the diff, then copies new file over the old one
4) It is fun

**Comment by sksamuel** [comment]
So a bit like snapshot testing.

**Comment by vlsi** [comment]
It is a snapshot testing.

**Comment by sksamuel** [comment]
Will look to include this for 5.1

**Comment by vlsi** [comment]
Do you intend to store expected values out-of-line or do you intend to store the values right in the regular test code?

**Comment by sksamuel** [comment]
I was thinking something like

foo shouldMatchSnapshot "/some/resource"

And if it fails, it writes the current value to /tmp/testname and includes
the path in the assertion message.

What do you think?

**Comment by vlsi** [comment]
External files are used quite often, and there's an established framework for that https://approvaltests.com/

---

I wonder if there's a way to write assertions in code.
Frankly speaking, I have not explored that idea in detail, however, what if

```kotlin
val foo = ...;
assertEquals(
    foo,
    """
    asdfasldkf;alsj f;laskjdf
    safda sdf asd f; lkajsd;flk as
    """
)
```

Then, there might be post-test processing that goes over assertEquals calls and replaces the values.

Of course, it won't be able to rewrite `assertEquals` into `assertStartsWith`.

**Comment by sksamuel** [comment]
I think this is most useful for things like json blobs right ?
Otherwise, you can use all the existing kotest matchers.

Like their example of arrays seems pointless when you can do
listOf(1,2,3) shouldContain 1

or something like that.

**Comment by vlsi** [comment]
>I think this is most useful for things like json blobs right ?

That depends.

1. It is super awesome for cases when outputs vary over time. For instance, HTML markup might evolve dramatically, so updating the expectations everywhere would be time consuming. For cases like SQL optimizer, and other optimizers, adding a new optimization might shuffle the outputs.
2. It allows to code tests like https://en.wikipedia.org/wiki/Trust_on_first_use. In other words, I type something, I launch it, I review the output, and accept it. For instance, the results of SQL might take time to format, however, if it prints the results, I can review it quite fast

A bit of a problem is parameterized tests, where "expected outputs" are typically passed as a parameter to the test method, so assertion methods do not touch expected value literal.

---

> Like their example of arrays seems pointless when you can do
> `listOf(1,2,3) shouldContain 1`

`shouldContain` might be useful indeed, however, I am not fond of those types of assertions because they are time-consuming to update in case the output changes 🤷 

In other words, I typically try to express the output as `assertEquals` on a text-like output.

**Comment by sksamuel** [comment]
I don't think we'd add support for arbitrary tests via this, but just
comparing string output.
The power of the kotest assertions module comes from the strong typing, so
things like array checks are not aligned with the library's goals.

Snapshot testing for html, json, other resources like that I think is great.

**Comment by sidgate** [comment]
Approval tests works well for API testing, I can just have the JSON verified. Now I cannot use kotest for such scenarios.

**Comment by nedtwigg** [comment]
- https://github.com/kotest/kotest/pull/3870
